README.md
# Physics-Definitions-Kivy-App
Kivy app to help practice and memorise H2 Physics definitions


changelog.txt
0.1 - base
0.1.1 - could not get it to work with kivymd(WORKED ON NATALIES PHONE), so removed it base but supports my phone (im not sure but might be the missing sdlttf that was causing problems)
0.2 - fix the list all not showing problem


underline_test.py

#function to help underline what i need later on
def underline(string):
    emptystring = ''

    for i in range(0, len(string)):

        if string[i] == ' ':
            emptystring = emptystring + string[i]
        else:
            emptystring= emptystring+string[i]+str('\u0332')

    return emptystring




icon_preview.py
from kivy.lang import Builder
from kivy.properties import StringProperty
from kivy.uix.screenmanager import Screen

from kivymd.icon_definitions import md_icons
from kivymd.app import MDApp
from kivymd.uix.list import OneLineIconListItem


Builder.load_string(
    '''
#:import images_path kivymd.images_path


<CustomOneLineIconListItem>

    IconLeftWidget:
        icon: root.icon


<PreviousMDIcons>

    MDBoxLayout:
        orientation: 'vertical'
        spacing: dp(10)
        padding: dp(20)

        MDBoxLayout:
            adaptive_height: True

            MDIconButton:
                icon: 'magnify'

            MDTextField:
                id: search_field
                hint_text: 'Search icon'
                on_text: root.set_list_md_icons(self.text, True)

        RecycleView:
            id: rv
            key_viewclass: 'viewclass'
            key_size: 'height'

            RecycleBoxLayout:
                padding: dp(10)
                default_size: None, dp(48)
                default_size_hint: 1, None
                size_hint_y: None
                height: self.minimum_height
                orientation: 'vertical'
'''
)


class CustomOneLineIconListItem(OneLineIconListItem):
    icon = StringProperty()


class PreviousMDIcons(Screen):

    def set_list_md_icons(self, text="", search=False):
        '''Builds a list of icons for the screen MDIcons.'''

        def add_icon_item(name_icon):
            self.ids.rv.data.append(
                {
                    "viewclass": "CustomOneLineIconListItem",
                    "icon": name_icon,
                    "text": name_icon,
                    "callback": lambda x: x,
                }
            )

        self.ids.rv.data = []
        for name_icon in md_icons.keys():
            if search:
                if text in name_icon:
                    add_icon_item(name_icon)
            else:
                add_icon_item(name_icon)


class MainApp(MDApp):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.screen = PreviousMDIcons()

    def build(self):
        return self.screen

    def on_start(self):
        self.screen.set_list_md_icons()


MainApp().run()


the programming.py
import random

#function to help underline what i need later on
def underline(string):
    emptystring = ''

    for i in range(0, len(string)):

        if string[i] == ' ':
            emptystring = emptystring + string[i]
        else:
            emptystring= emptystring+string[i]+str('\u0332')

    return emptystring


#the actual stuff

#read the definition file
definitions = {}
with open('definitions.txt') as f1:
    #each line should have 1 definition, tab seperated
    for line in f1:
        word, definition = line.strip().split('\t')
        definitions[word] = definition

#repeatedly ask user what they want
quit_flag = False
one_or_two = False #if i choose option one or two
while quit_flag == False:
    option = input('''
Please choose an option:
1. Enter a specific word (that you would like to define)
2. Random
3. Quit
''')
    if option == '1':
        one_or_two = True
        exist = False
        while exist == False:
            word = input('Please enter a word/phrase you would like to define: ')
            word = word.lower()
            if word in definitions:
                exist = True
            else:
                print('That word/phrase does not exist! Did you spell it right?')

        actual_definition = definitions[word]
        
    elif option == '2':
        one_or_two = True
        #choose a random pair
        word, actual_definition = random.choice(list(definitions.items())) 

    elif option =='3':
        print('Goodbye!')
        break
    
    else:
        print('That was neither 1, 2, or 3!')

    #So i dont need to repeat code
    #this is to enter the defnition and compare it with the choosen word
    #to get here should have word and defnition variables already
    if one_or_two == True:
        #ask user toenter their definition
        user_answer = input(f'Please define {word}: ')

        #get each individual word, so i can see which words i missed
        actual_definition_words = actual_definition.strip().split(' ') #this is for compare
        output_definition_words = actual_definition_words.copy() #this is my backup copy
        user_answer_words = user_answer.strip().split(' ')

        #compare
        for word in user_answer_words:
            word = word.lower()
            if (word in actual_definition_words):
                actual_definition_words.remove(word) #remove the word

            #sometimes the word is connected to the comma
            elif (word+',') in actual_definition_words:
                actual_definition_words.remove(word+',')
    
        #whatever is left in actual_definition_words would be the once i missed
        print('Here is the full definition. The underlined words are the words you missed:') 

        output = ''

        for word in output_definition_words:
            #if its in that list, means did not mention, so underline, somecommonuseless words
            if (word in actual_definition_words) and (word not in ['the', 'of', 'is', 'it']):
                word = underline(word)

            output += word + ' '

        print(output)


0.1/phy_def_helper_v0.1.py
from kivymd.app import MDApp
from kivy.lang.builder import Builder
from kivy.uix.screenmanager import Screen, ScreenManager
from kivy.uix.label import Label
from kivy.uix.scrollview import ScrollView
import threading
import random

Builder_string = '''
ScreenManager:
    Main:
<Main>:
    name : 'Definition App'
    BoxLayout:
        orientation: "vertical"
        MDToolbar:
            title: 'Definition App'
            md_bg_color: app.theme_cls.primary_color
            specific_text_color: 1, 1, 1, 1
        MDTextField:
            id: your_query
            hint_text: "Your Query (The word you want to define)"
            color_mode: 'custom'
            helper_text_mode: "on_focus"
            markup: True
        MDTextField:
            id: your_definition
            hint_text: "Your Definition"
            color_mode: 'custom'
            helper_text_mode: "on_focus"
            markup: True

        ScrollView:
            Label:
                id: label_1
                text: 'Actual Definition (Missing words will be bolded and underlined)'
                color:'#1675f7'
                markup: True
                multiline: True
                size_hint_y: None
                size: self.texture_size
                height: self.texture_size[1]  
                text_size: self.width, None
                

        MDRectangleFlatIconButton:
            text: "Check"
            icon: "clipboard-search"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            on_press: app.run_fun()
        MDRectangleFlatIconButton:
            text: "Random"
            icon: "boomerang"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            on_press: app.random()
        MDRectangleFlatIconButton:
            text: "List All"
            icon: "view-list"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            on_press: app.list_all()
            
        MDSpinner:
            id: rc_spin
            size_hint: None, None
            size: dp(46), dp(46)
            pos_hint: {'center_x': .5, 'center_y': .5}
            active: False
    '''


class Main(Screen):
    pass


sm = ScreenManager()
sm.add_widget(Main(name='Definition_App'))


class MainApp(MDApp):
    def build(self):
        self.help_string = Builder.load_string(Builder_string)
        self.title = 'Definition App'
        self.definitions = {}
        with open('physics_definitions.txt') as f1:
            #each line should have 1 definition, tab seperated
            for line in f1:
                word, definition = line.strip().split('\t')
                self.definitions[word] = definition

        return self.help_string

    def run_fun(self):
        t1 = threading.Thread(target=self.search)
        t1.start()

    def random(self):
        random_query, definition = random.choice(list(self.definitions.items()))
        self.help_string.get_screen('Definition App').ids.your_query.text = random_query

    def list_all(self):
        full_text = ''
        self.help_string.get_screen('Definition App').ids.rc_spin.active = True
        for word in self.definitions:
            definition = self.definitions[word]
            full_text += f'[b]{word}[/b]:  {definition}\n\n'
            
        self.help_string.get_screen('Definition App').ids.label_1.text = full_text
        self.help_string.get_screen('Definition App').ids.rc_spin.active = False    
        
    def search(self):
        self.help_string.get_screen('Definition App').ids.rc_spin.active = True
        query = self.help_string.get_screen('Definition App').ids.your_query.text
        user_definition = self.help_string.get_screen('Definition App').ids.your_definition.text
        result = self.get_result(query, user_definition)
##        result = wikipedia.summary(query, sentences=6)
        result = result.strip()
        self.help_string.get_screen('Definition App').ids.label_1.text = result
        self.help_string.get_screen('Definition App').ids.rc_spin.active = False

    #function to help underline words i missed later on
    def underline(self, string):
##        emptystring = ''
##
##        for i in range(0, len(string)):
##
##            if string[i] == ' ':
##                emptystring = emptystring + string[i]
##            else:
##                emptystring= emptystring+string[i]+str('̲') #'\u0332'
##
##        #return emptystring

        #have to redefine cause kivy dont know underline unicode
        return '[b][u]'+string+'[/u][/b]'

    def get_result(self, query, user_definition):
        exist = False
        #get actual definition
        query = query.lower()
        if query in self.definitions:
            exist = True
            actual_definition = self.definitions[query]
        else:
            return 'I do not know the definition to the word you tried to define, did you spell it correctly? Or try phrasing it differently.'

        if exist == True:
            #get each individual word, so i can see which words i missed
            actual_definition_words = actual_definition.strip().split(' ') #this is for compare
            output_definition_words = actual_definition_words.copy() #this is my backup copy
            user_answer_words = user_definition.strip().split(' ')

            #compare
            for word in user_answer_words:
                word = word.lower()
                if (word in actual_definition_words):
                    actual_definition_words.remove(word) #remove the word

                #sometimes the word is connected to the comma, full stop, or without it
                elif (word+',') in actual_definition_words:
                    actual_definition_words.remove(word+',')

                elif (word+'.') in actual_definition_words:
                    actual_definition_words.remove(word+'.')

                elif (word[:-1]) in actual_definition_words:
                    actual_definition_words.remove(word[:-1])



            output = ''

            for word in output_definition_words:
                #if its in that list, means did not mention, so underline
                if (word in actual_definition_words) and (word not in ['an', 'the', 'of', 'is', 'it', 'a', 'to', 'at', 'are', 'that']):
                    word = self.underline(word)

                output += word + ' '
            return output

    
MainApp().run()


0.1/phy_def_helper_v0.1_WITHOUT_KIVYMD.py
from kivy.app import App
from kivy.lang.builder import Builder
from kivy.uix.screenmanager import Screen, ScreenManager
from kivy.uix.label import Label
from kivy.uix.scrollview import ScrollView
from kivy.uix.textinput import TextInput
from kivy.uix.button import Button
import threading
import random

Builder_string = '''
ScreenManager:
    Main:
<Main>:
    name : 'Definition App'
    BoxLayout:
        orientation: "vertical"

        TextInput:
            id: your_query
            hint_text: "Your Query (The word you want to define)"
            color_mode: 'custom'
            helper_text_mode: "on_focus"
            size_hint_y: 0.2
            markup: True

        TextInput:
            id: your_definition
            hint_text: "Your Definition"
            color_mode: 'custom'
            helper_text_mode: "on_focus"
            size_hint_y: 0.2
            markup: True

        ScrollView:
            Label:
                id: label_1
                text: 'Actual Definition (Missing words will be bolded and underlined)'
                markup: True
                multiline: True
                size_hint_y: None
                size: self.texture_size
                height: self.texture_size[1]  
                text_size: self.width, None
                font_size: '15sp'
                

        Button:
            text: "Check"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            size_hint_y: 0.15
            on_press: app.run_fun()
        Button:
            text: "Random"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            size_hint_y: 0.15
            on_press: app.random()
        Button:
            text: "List All"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            size_hint_y: 0.15
            on_press: app.list_all()

    '''


class Main(Screen):
    pass


sm = ScreenManager()
sm.add_widget(Main(name='Definition_App'))


class MainApp(App):
    def build(self):
        self.help_string = Builder.load_string(Builder_string)
        self.title = 'Definition App'
        self.definitions = {}
        with open('physics_definitions.txt') as f1:
            #each line should have 1 definition, tab seperated
            for line in f1:
                word, definition = line.strip().split('\t')
                self.definitions[word] = definition

        return self.help_string

    def run_fun(self):
        t1 = threading.Thread(target=self.search)
        t1.start()

    def random(self):
        random_query, definition = random.choice(list(self.definitions.items()))
        self.help_string.get_screen('Definition App').ids.your_query.text = random_query

    def list_all(self):
        full_text = ''
        for word in self.definitions:
            definition = self.definitions[word]
            full_text += f'[b]{word}[/b]:  {definition}\n\n'
            
        self.help_string.get_screen('Definition App').ids.label_1.text = full_text    
        
    def search(self):
        query = self.help_string.get_screen('Definition App').ids.your_query.text
        user_definition = self.help_string.get_screen('Definition App').ids.your_definition.text
        result = self.get_result(query, user_definition)
##        result = wikipedia.summary(query, sentences=6)
        result = result.strip()
        self.help_string.get_screen('Definition App').ids.label_1.text = result

    #function to help underline words i missed later on
    def underline(self, string):
##        emptystring = ''
##
##        for i in range(0, len(string)):
##
##            if string[i] == ' ':
##                emptystring = emptystring + string[i]
##            else:
##                emptystring= emptystring+string[i]+str('̲') #'\u0332'
##
##        #return emptystring

        #have to redefine cause kivy dont know underline unicode
        return '[b][u]'+string+'[/u][/b]'

    def get_result(self, query, user_definition):
        exist = False
        #get actual definition
        query = query.lower()
        if query in self.definitions:
            exist = True
            actual_definition = self.definitions[query]
        else:
            return 'I do not know the definition to the word you tried to define, did you spell it correctly? Or try phrasing it differently.'

        if exist == True:
            #get each individual word, so i can see which words i missed
            actual_definition_words = actual_definition.strip().split(' ') #this is for compare
            output_definition_words = actual_definition_words.copy() #this is my backup copy
            user_answer_words = user_definition.strip().split(' ')

            #compare
            for word in user_answer_words:
                word = word.lower()
                if (word in actual_definition_words):
                    actual_definition_words.remove(word) #remove the word

                #sometimes the word is connected to the comma, full stop, or without it
                elif (word+',') in actual_definition_words:
                    actual_definition_words.remove(word+',')

                elif (word+'.') in actual_definition_words:
                    actual_definition_words.remove(word+'.')

                elif (word[:-1]) in actual_definition_words:
                    actual_definition_words.remove(word[:-1])



            output = ''

            for word in output_definition_words:
                #if its in that list, means did not mention, so underline
                if (word in actual_definition_words) and (word not in ['an', 'the', 'of', 'is', 'it', 'a', 'to', 'at', 'are', 'that']):
                    word = self.underline(word)

                output += word + ' '
            return output

    
MainApp().run()


0.2/phy_def_helper_v0.2_WITHOUT_KIVYMD.py
from kivy.app import App
from kivy.lang.builder import Builder
from kivy.uix.screenmanager import Screen, ScreenManager
from kivy.uix.label import Label
from kivy.uix.scrollview import ScrollView
from kivy.uix.textinput import TextInput
from kivy.uix.button import Button
import threading
import random

Builder_string = '''
ScreenManager:
    Main:
<Main>:
    name : 'Definition App'
    BoxLayout:
        orientation: "vertical"

        TextInput:
            id: your_query
            hint_text: "Your Query (The word you want to define)"
            color_mode: 'custom'
            helper_text_mode: "on_focus"
            size_hint_y: 0.2
            markup: True

        TextInput:
            id: your_definition
            hint_text: "Your Definition"
            color_mode: 'custom'
            helper_text_mode: "on_focus"
            size_hint_y: 0.2
            markup: True

        ScrollView:
            Label:
                id: label_1
                text: 'Actual Definition (Missing words will be bolded and underlined)'
                markup: True
                multiline: True
                size_hint_y: None
                size: self.texture_size
                height: self.texture_size[1]  
                text_size: self.width, None
                font_size: '15sp'
        ScrollView:
            Label:
                id: label_2
                text: ''
                markup: True
                multiline: True
                size_hint_y: None
                size: self.texture_size
                height: self.texture_size[1]  
                text_size: self.width, None
                font_size: '15sp'
                

        Button:
            text: "Check"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            size_hint_y: 0.15
            on_press: app.run_fun()
        Button:
            text: "Random"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            size_hint_y: 0.15
            on_press: app.random()
        Button:
            text: "List All"
            line_color: 0, 0, 0, 0
            pos_hint: {"center_x": .5, "center_y": .6}
            size_hint_y: 0.15
            on_press: app.list_all()

    '''


class Main(Screen):
    pass


sm = ScreenManager()
sm.add_widget(Main(name='Definition_App'))


class MainApp(App):
    def build(self):
        self.help_string = Builder.load_string(Builder_string)
        self.title = 'Definition App'
        self.definitions = {}
        with open('physics_definitions.txt') as f1:
            #each line should have 1 definition, tab seperated
            for line in f1:
                word, definition = line.strip().split('\t')
                self.definitions[word] = definition

        return self.help_string

    def run_fun(self):
        t1 = threading.Thread(target=self.search)
        t1.start()

    def random(self):
        random_query, definition = random.choice(list(self.definitions.items()))
        self.help_string.get_screen('Definition App').ids.your_query.text = random_query
        self.help_string.get_screen('Definition App').ids.label_1.text = 'Actual Definition (Missing words will be bolded and underlined)'
        self.help_string.get_screen('Definition App').ids.label_2.text = ''
        
        
    def list_all(self):
        first_half = ''
        second_half = ''
        num_of_definitions = len(self.definitions)
        half = num_of_definitions//2
        counter = 0
        #to try address the label not showeing the full text, split into 2 labels    
        for word in self.definitions:
            definition = self.definitions[word]
            if counter < half:
                first_half += f'[b]{word}[/b]:  {definition}\n\n'
            else:
                second_half += f'[b]{word}[/b]:  {definition}\n\n'

            counter += 1
            
        self.help_string.get_screen('Definition App').ids.label_1.text = first_half
        self.help_string.get_screen('Definition App').ids.label_2.text = second_half
        
        
    def search(self):
        query = self.help_string.get_screen('Definition App').ids.your_query.text
        user_definition = self.help_string.get_screen('Definition App').ids.your_definition.text
        result = self.get_result(query, user_definition)
##        result = wikipedia.summary(query, sentences=6)
        result = result.strip()
        self.help_string.get_screen('Definition App').ids.label_1.text = result
        self.help_string.get_screen('Definition App').ids.label_2.text = ''
        

    #function to help underline words i missed later on
    def underline(self, string):
##        emptystring = ''
##
##        for i in range(0, len(string)):
##
##            if string[i] == ' ':
##                emptystring = emptystring + string[i]
##            else:
##                emptystring= emptystring+string[i]+str('̲') #'\u0332'
##
##        #return emptystring

        #have to redefine cause kivy dont know underline unicode
        return '[b][u]'+string+'[/u][/b]'

    def get_result(self, query, user_definition):
        exist = False
        #get actual definition
        query = query.lower()
        if query in self.definitions:
            exist = True
            actual_definition = self.definitions[query]
        else:
            return 'I do not know the definition to the word you tried to define, did you spell it correctly? Or try phrasing it differently.'

        if exist == True:
            #get each individual word, so i can see which words i missed
            actual_definition_words = actual_definition.strip().split(' ') #this is for compare
            output_definition_words = actual_definition_words.copy() #this is my backup copy
            user_answer_words = user_definition.strip().split(' ')

            #compare
            for word in user_answer_words:
                word = word.lower()
                if (word in actual_definition_words):
                    actual_definition_words.remove(word) #remove the word

                #sometimes the word is connected to the comma, full stop, or without it
                elif (word+',') in actual_definition_words:
                    actual_definition_words.remove(word+',')

                elif (word+'.') in actual_definition_words:
                    actual_definition_words.remove(word+'.')

                elif (word[:-1]) in actual_definition_words:
                    actual_definition_words.remove(word[:-1])



            output = ''

            for word in output_definition_words:
                #if its in that list, means did not mention, so underline
                if (word in actual_definition_words) and (word not in ['an', 'the', 'of', 'is', 'it', 'a', 'to', 'at', 'are', 'that']):
                    word = self.underline(word)

                output += word + ' '
            return output

    
MainApp().run()
